### ISSUE TITLE
[RLlib][Bug] Wrong number of samples in `batch[SampleBatch.INFOS]` in RNN cases

### ISSUE THREAD

**Issue by XuehaiPan** [ray-project/ray#21991]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I'm trying to train my custom models by adding additional custom loss with `info`. This is a common case for training RL policies with auxiliary tasks and online imitation with behavior cloning.

I add some auxiliary information in my environment's `step` method. For example, the field `info` contains the component of the reward function.

```python
class MyEnv(gym.Env):
    def step(self, action):
        # forward dynamic
        reward_1 = reward_func_1()
        reward_2 = reward_func_2()
        reward = reward_1 + reward_2

        info = {'reward_1': reward_1, 'reward_2': reward_2}

        return observation, reward, done, info
```

My custom model with custom loss:

```python
class CustomModel(TorchRNN):
    def custom_loss(self, policy_loss, loss_inputs):

        seq_lens = loss_inputs[SampleBatch.SEQ_LENS]
        infos = loss_inputs[SampleBatch.INFOS]

        B = len(loss_inputs[SampleBatch.SEQ_LENS])
        num_padded_samples = len(loss_inputs[SampleBatch.OBS])

        max_seq_len = num_padded_samples // B
        mask = sequence_mask(seq_lens, maxlen=max_seq_len)

        reward_1 = ... # extract from `loss_inputs[SampleBatch.INFOS]`
        reward_2 = ... # extract from `loss_inputs[SampleBatch.INFOS]`

        reward_1 = ... # padding and add time dimension
        reward_2 = ... # padding and add time dimension

        reward_1_predicted = self.reward_1_prediction_branch(self._features)
        reward_2_predicted = self.reward_2_prediction_branch(self._features)

        custom_loss = ... # reward prediction loss using items from `info`

        return policy_loss + custom_loss
```

In RNN cases, the sample batch will be padded and stacked into `torch.Tensor` before feeding into method `model.forward` and `model.custom_loss`, except `INFOS` and `SEQ_LENS`:

https://github.com/ray-project/ray/blob/b73a007ccde13d809b20252478f086b4ad997e72/rllib/policy/rnn_sequencing.py#L109-L120

For example, if I get a batch with batch size 16 and max sequence length 25, then:

```python
# after padding
batch[SampleBatch.OBS]       # a torch tensor (dtype: float,  shape: (16, 25, *))
batch[SampleBatch.SEQ_LENS]  # a torch tensor (dtype: int,    shape: (16,))
batch[SampleBatch.INFOS]     # a numpy array  (dtype: object, shape: (num_info,))
```

Although for compatibility reasons, RLlib cannot assume the content in `info`. I can make my own batch from `info` by manual padding and reshaping. In sample batch, it should have:

```python
num_infos == seq_lens.sum()
```

### Expected

Ideally:

```python
len(batch[SampleBatch.INFOS]) == batch[SampleBatch.SEQ_LENS].sum().item()
```

I don't really sure about the details of how RLlib construct sample batch for RNNs. Both `rollout_fregment_length` and `max_seq_len` in `config` will affect the input batch. So I think the following is also acceptable:

```python
len(batch[SampleBatch.INFOS]) >= batch[SampleBatch.SEQ_LENS].sum().item()
```

### What Happened

Sometimes (not always), I got:

```python
len(batch[SampleBatch.INFOS]) < batch[SampleBatch.SEQ_LENS].sum().item()
```

This usually happens when `num_vaild_samples != B * T`, i.e., the episode is complete and the batch is padded.

![image](https://docs.ray.io/en/latest/_images/rllib-batch-modes.svg)

I can reproduce this issue with either `"batch_mode": "truncate_episodes"` or `"batch_mode": "complete_episodes"`.

### Versions / Dependencies

Ray nightly

### Reproduction script

```python
import argparse
import os
import sys
from warnings import warn

import ray
from ray import tune
from ray.rllib.examples.models.rnn_model import TorchRNNModel
from ray.rllib.models import ModelCatalog
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.utils.torch_utils import sequence_mask


DEBUG = (getattr(sys, 'gettrace', lambda: None)() is not None)


class TorchRNNModelWithInfos(TorchRNNModel):
    def custom_loss(self, policy_loss, loss_inputs):

        seq_lens = loss_inputs[SampleBatch.SEQ_LENS]
        infos = loss_inputs[SampleBatch.INFOS]

        B = len(loss_inputs[SampleBatch.SEQ_LENS])
        num_padded_samples = len(loss_inputs[SampleBatch.OBS])

        max_seq_len = num_padded_samples // B
        mask = sequence_mask(seq_lens, maxlen=max_seq_len)

        num_valid_samples = mask.sum().item()
        num_infos = len(infos)

        assert len(loss_inputs) == num_valid_samples
        assert seq_lens.sum().item() == num_valid_samples

        if num_infos < num_valid_samples:
            warn(f"Insufficient number of `info` "
                 f"(expected {num_valid_samples}, got {len(infos)}).")
        elif num_infos > num_valid_samples:
            warn(f"Incorrect number of `info` "
                 f"(expected {num_valid_samples}, got {len(infos)}).")

        # assert num_infos >= num_valid_samples

        return policy_loss


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--run", type=str, default="PPO", help="The RLlib-registered algorithm to use."
    )
    parser.add_argument("--env", type=str, default="CartPole-v0")
    parser.add_argument("--num-cpus", type=int, default=0)
    parser.add_argument(
        "--stop-iters", type=int, default=100, help="Number of iterations to train."
    )
    parser.add_argument(
        "--stop-timesteps", type=int, default=100000, help="Number of timesteps to train."
    )

    args = parser.parse_args()

    ray.init(num_cpus=args.num_cpus or None, local_mode=DEBUG)

    ModelCatalog.register_custom_model(
        "rnn", TorchRNNModelWithInfos
    )

    config = {
        "env": args.env,
        "framework": "torch",
        "gamma": 0.9,
        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        "_fake_gpus": not DEBUG,
        "num_workers": 0,
        "num_envs_per_worker": 20,
        "entropy_coeff": 0.001,
        "num_sgd_iter": 5,
        "vf_loss_coeff": 1E-5,
        "model": {
            "custom_model": "rnn",
            "max_seq_len": 20,
            "custom_model_config": {
                "cell_size": 32,
            },
        },

        # "batch_mode": "complete_episodes",
        "batch_mode": "truncate_episodes",
    }

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
    }

    results = tune.run(args.run, config=config, stop=stop, verbose=1)

    ray.shutdown()


if __name__ == "__main__":
    main()
```

I got:

```
(PPOTrainer pid=314050) 2022-01-31 16:48:27,935 WARNING util.py:60 -- Install gputil for GPU system monitoring.
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 130, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 122, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 132, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 125, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 119, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 119, got 100).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 137, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 139, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "

(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 112, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 141, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 136, got 40).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 131, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 117, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 124, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "

(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 131, got 100).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 134, got 80).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 135, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 113, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
```

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!
